File: library/src/main/java/com/owncloud/android/lib/resources/albums/RenameAlbumRemoteOperation.java
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2025 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package com.owncloud.android.lib.resources.albums;

import com.nextcloud.common.SessionTimeOut;
import com.nextcloud.common.SessionTimeOutKt;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.network.WebdavUtils;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;

import org.apache.jackrabbit.webdav.client.methods.MoveMethod;

public class RenameAlbumRemoteOperation extends RemoteOperation {
    private static final String TAG = RenameAlbumRemoteOperation.class.getSimpleName();
    private final String mOldRemotePath;
    private final String mNewAlbumName;
    private final SessionTimeOut sessionTimeOut;

    public RenameAlbumRemoteOperation(String mOldRemotePath, String mNewAlbumName) {
        this(mOldRemotePath, mNewAlbumName, SessionTimeOutKt.getDefaultSessionTimeOut());
    }

    public RenameAlbumRemoteOperation(String mOldRemotePath, String mNewAlbumName, SessionTimeOut sessionTimeOut) {
        this.mOldRemotePath = mOldRemotePath;
        this.mNewAlbumName = mNewAlbumName;
        this.sessionTimeOut = sessionTimeOut;
    }

    public String getNewAlbumName() {
        return mNewAlbumName;
    }

    /**
     * Performs the operation.
     *
     * @param client Client object to communicate with the remote ownCloud server.
     */
    @Override
    protected RemoteOperationResult run(OwnCloudClient client) {
        RemoteOperationResult result = null;
        MoveMethod move = null;
        String url = client.getBaseUri() + "/remote.php/dav/photos/" + client.getUserId() + "/albums";
        try {
            if (!this.mNewAlbumName.equals(this.mOldRemotePath)) {
                move = new MoveMethod(url + WebdavUtils.encodePath(mOldRemotePath), url + WebdavUtils.encodePath(mNewAlbumName), true);
                client.executeMethod(move, sessionTimeOut.getReadTimeOut(), sessionTimeOut.getConnectionTimeOut());
                result = new RemoteOperationResult(move.succeeded(), move);
                Log_OC.i(TAG, "Rename " + this.mOldRemotePath + " to " + this.mNewAlbumName + ": " + result.getLogMessage());
                client.exhaustResponse(move.getResponseBodyAsStream());
                return result;
            }
        } catch (Exception e) {
            result = new RemoteOperationResult(e);
            Log_OC.e(TAG, "Rename " + this.mOldRemotePath + " to " + this.mNewAlbumName + ": " + result.getLogMessage(), e);
            return result;
        } finally {
            if (move != null) {
                move.releaseConnection();
            }
        }

        return result;
    }

}

File: library/src/main/java/com/owncloud/android/lib/resources/albums/RenameAlbumRemoteOperation.kt
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2026 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.albums

import com.nextcloud.common.SessionTimeOut
import com.nextcloud.common.defaultSessionTimeOut
import com.owncloud.android.lib.common.OwnCloudClient
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.operations.RemoteOperationResult.ResultCode
import com.owncloud.android.lib.common.utils.Log_OC
import org.apache.commons.httpclient.HttpStatus
import org.apache.jackrabbit.webdav.client.methods.MoveMethod

class RenameAlbumRemoteOperation
    @JvmOverloads
    constructor(
        private val oldRemotePath: String,
        val newAlbumName: String,
        private val sessionTimeOut: SessionTimeOut = defaultSessionTimeOut
    ) : RemoteOperation<Any>() {
        @Deprecated("Deprecated in Java")
        @Suppress("TooGenericExceptionCaught", "DEPRECATION")
        override fun run(client: OwnCloudClient): RemoteOperationResult<Any> {
            if (newAlbumName == oldRemotePath) {
                return RemoteOperationResult(ResultCode.OK)
            }

            var move: MoveMethod? = null
            return try {
                move = MoveMethod(client.albumUri(oldRemotePath), client.albumUri(newAlbumName), false)
                client.executeMethod(move, sessionTimeOut.readTimeOut, sessionTimeOut.connectionTimeOut)

                val renamed = RemoteOperationResult<Any>(move.succeeded(), move)
                Log_OC.i(TAG, "Rename $oldRemotePath to $newAlbumName : ${renamed.logMessage}")

                val result =
                    if (renamed.httpCode == HttpStatus.SC_PRECONDITION_FAILED) {
                        // the target album name is already taken
                        RemoteOperationResult(ResultCode.INVALID_OVERWRITE)
                    } else {
                        renamed
                    }
                client.exhaustResponse(move.responseBodyAsStream)
                result
            } catch (e: Exception) {
                failure(e)
            } finally {
                move?.releaseConnection()
            }
        }

        @Suppress("DEPRECATION")
        private fun failure(e: Exception): RemoteOperationResult<Any> =
            RemoteOperationResult<Any>(e).also {
                Log_OC.e(TAG, "Rename $oldRemotePath to $newAlbumName : ${it.logMessage}", e)
            }

        companion object {
            private val TAG: String = RenameAlbumRemoteOperation::class.java.simpleName
        }
    }
